job_creator.php

```diff
@@ -75,6 +79,51 @@ public function createJob(int $phpIndex, array $opts): array
         return array_merge($default, $opts);
     }
 
+    private function isAllowedPhpVersion(string $phpVersion)
+    {
+        $phpVersion = (float) $phpVersion;
+        // no composer.json file or php version no defined in composer.json, just allow the php version
+        if ($this->composerPhpConstraint === null) {
+            return true;
+        }
+        if (!file_exists($this->composerJsonPath)) {
+            $this->composerPhpConstraint = null;
+            return true;
+        }
+        $json = json_decode(file_get_contents($this->composerJsonPath));
+        if (!isset($json->require->php)) {
+            $this->composerPhpConstraint = null;
+            return true;
+        }
+        if ($this->composerPhpConstraint === '') {
+            $this->composerPhpConstraint = $json->require->php;
+        }
```

Comment:
```suggestion
        if ($this->composerPhpConstraint === '') {
            if (!file_exists($this->composerJsonPath)) {
                $this->composerPhpConstraint = null;
                return true;
            }
            $json = json_decode(file_get_contents($this->composerJsonPath));
            if (!isset($json->require->php)) {
                $this->composerPhpConstraint = null;
                return true;
            }
            $this->composerPhpConstraint = $json->require->php;
        }
```
We don't need to do all of those checks every time - just the first time. Possibly this should even be its own separate method but I'm okay with it staying in here.

Reply:
> We don't need to do all of those checks every time

That's right, there's a check above this

```php
    if ($this->composerPhpConstraint === null) {
        return true;
     }
```

Reply:
`$this->composerPhpConstraint` is a `?string`, not a bool at all, but that's not the point.

Currently if `$this->composerPhpConstraint` holds some valid constraint string, we _still_ go through all of the above checks (does the composer.json file exist? Is there a php constraint in the composer.json file?) which is unnecessary, because by this point we already have the constraint.

Reply:
updated